sdk/core/azure-core/azure/core/rest/_http_response_impl.py

```diff
@@ -0,0 +1,321 @@
+        self._content = None  # type: Optional[bytes]
+        self._text = None  # type: Optional[str]
+
+    @property
+    def request(self):
+        # type: (...) -> _HttpRequest
+        """The request that resulted in this response.
+
+        :rtype: ~azure.core.rest.HttpRequest
+        """
+        return self._request
+
+    @property
+    def url(self):
+        # type: (...) -> str
+        """The URL that resulted in this response.
+
+        :rtype: str
+        """
+        return self.request.url
+
+    @property
+    def is_closed(self):
+        # type: (...) -> bool
+        """Whether the network connection has been closed yet.
+
+        :rtype: bool
+        """
+        return self._is_closed
+
+    @property
+    def is_stream_consumed(self):
+        # type: (...) -> bool
+        """Whether the stream has been fully consumed"""
+        return self._is_stream_consumed
+
+    @property
+    def status_code(self):
+        # type: (...) -> int
+        """The status code of this response.
+
+        :rtype: int
+        """
+        return self._status_code
+
+    @property
+    def headers(self):
+        # type: (...) -> MutableMapping[str, str]
+        """The response headers.
+
+        :rtype: MutableMapping[str, str]
+        """
+        return self._headers
+
+    @property
+    def content_type(self):
+        # type: (...) -> Optional[str]
+        """The content type of the response.
+
+        :rtype: optional[str]
+        """
+        return self._content_type
+
+    @property
+    def reason(self):
+        # type: (...) -> str
+        """The reason phrase for this response.
+
+        :rtype: str
+        """
+        return self._reason
+
+    @property
+    def encoding(self):
+        # type: (...) -> Optional[str]
+        """Returns the response encoding.
+
+        :return: The response encoding. We either return the encoding set by the user,
+         or try extracting the encoding from the response's content type. If all fails,
+         we return `None`.
+        :rtype: optional[str]
+        """
+        try:
+            return self._encoding
+        except AttributeError:
+            self._encoding = get_charset_encoding(self)  # type: Optional[str]
+            return self._encoding
+
+    @encoding.setter
+    def encoding(self, value):
+        # type: (str) -> None
+        """Sets the response encoding"""
+        self._encoding = value
+        self._text = None  # clear text cache
+        self._json = None  # clear json cache as well
+
+    def text(self, encoding=None):
+        # type: (Optional[str]) -> str
+        """Returns the response body as a string
+
+        :param optional[str] encoding: The encoding you want to decode the text with. Can
+         also be set independently through our encoding property
+        :return: The response's content decoded as a string.
+        """
+        if self._text is None or encoding:
```

Comment:
is it by design that

```py
text1 = response.text()
text2 = response.text(encode="asii")
text3 = response.text()

text1 != text3
```
?

Reply:
no that's not by design, let me fix this and add a test. thanks!

Reply:
Then should we have def json(self, encoding=None)?

Reply:
resolved offline with @xiangyan99 !